lib/mllparty/connection_hub/client_monitor.ex

```diff
@@ -11,91 +10,129 @@ defmodule MLLParty.ConnectionHub.ClientMonitor do
 
   alias MLLParty.ConnectionHub.ClientWrapper
 
-  @impl true
-  def init(_opts), do: {:ok, %{
-    down_at: nil,
-    down_notification_sent: false
-  }}
+  def child_spec(ip, port, client_wrapper_pid) do
+    %{
+      id: "#{ip}:#{port} monitor",
+      start: {__MODULE__, :start_link, [ip, port, client_wrapper_pid]}
+    }
+  end
 
-  def start_link() do
-    GenServer.start_link(__MODULE__, [], name: @name)
+  @impl true
+  def init([ip, port, client_wrapper_pid]) do
+    Logger.info("Starting client monitor: #{ip}:#{port}")
+    schedule_monitor()
+    # Return :ok and initial state
+    {:ok, %{
+      ip: ip,
+      port: port,
+      client_wrapper_pid: client_wrapper_pid,
+      down_at: nil,
+      down_notification_sent: false,
+      reconnect_attempted: false
+    }}
   end
 
-  def start_monitor(client_wrapper_pid) do
-    %{endpoint: endpoint} = ClientWrapper.client_status(client_wrapper_pid)
-    Logger.info("Monitoring connection for #{endpoint}")
-    {:ok, pid} = start_link()
-    Process.send(pid, {:monitor, client_wrapper_pid}, [])
+  def start_link(ip, port, client_wrapper_pid) do
+    GenServer.start_link(__MODULE__, [ip, port, client_wrapper_pid], name: String.to_atom("#{ip}:#{port} monitor"))
   end
 
   @impl true
-  def handle_info({:monitor, client_wrapper_pid}, state) do
+  def handle_info(:monitor, state) do
     %{
       endpoint: endpoint,
       connected: connected,
       pending_reconnect: pending_reconnect
-    } = ClientWrapper.client_status(client_wrapper_pid)
-    Logger.info("Connected: #{connected}")
-    client_pid = ClientWrapper.child_client_pid(client_wrapper_pid)
-    # Get current timestamp
-    current_timestamp = timestamp()
+    } = ClientWrapper.client_status(state.client_wrapper_pid)
     # Check the connection status
-    {down_at, down_notification_sent} = case connected do
+    new_state = case connected == true do
```

Comment:
Nice refactor here– one nitpick is that a simple `if` is probably a bit more readable (also comparison with `true` is unnecessary):
```
new_state =
  if connected do
    handle_connection_up(endpoint, state)
  else
    handle_connection_down(endpoint, connected, pending_reconnect, state)
  end